modules/core/src/main/java/org/apache/ignite/internal/processors/service/DynamicServiceChangeRequest.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.service;
+
+import java.io.Serializable;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.services.ServiceConfiguration;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Service change request.
+ */
+public class DynamicServiceChangeRequest implements Serializable {
```

Comment:
Why don't we make separate classes for deploy and undeploy actions?
The flags thing seems over-complicated.

Reply:
I want to avoid creating multiple classes with similar fields, which will be processed in a unified flow. The main reason to maintain messages compatibility between releases. Also, we can use this class in the future for adding a new action "redeploy" by the new flag.

The flags give us the flexibility to add actions across versions without breaking serialization/deserialization.

BTW, this approach uses widely in Ignite and should be well known for the community, see `DynamicCacheChangeRequest`, `GridMemcachedMessage`, `GridDhtPartitionsAbstractMessage` etc.

Reply:
@dmekhanikov 

Actually, we take main principles from cache creation flow.
As you can see `DynamicCacheChangeRequest` shares similar struscture.

Reply:
The only common property you need to share between `deploy` and `undeploy` actions is `serviceId`. You can make an abstract class with this property.
`cfg` field is always `null` for the undeployment requests, so it doesn't need to be shared. And flags `deploy` and `undeploy` look like a bad design, because they can be set to true at the same time.
I don't see, how adding new message types violate compatibility. If you need to add a new action, then you'll just add another message type.

Reply:
Reworked.